File: examples/variance_reduction.cpp
#include "derivkit/derivkit.hpp"

#include <cmath>
#include <iomanip>
#include <iostream>

namespace {

void row(const char* name, const derivkit::PricingResult& r, double ref_err) {
    const double ratio = ref_err > 0.0 ? (ref_err * ref_err) / (r.error_estimate * r.error_estimate)
                                       : 0.0;
    std::cout << std::left << std::setw(28) << name << std::right << std::fixed
              << std::setprecision(6) << std::setw(12) << r.value << std::scientific
              << std::setprecision(3) << std::setw(12) << r.error_estimate << std::fixed
              << std::setprecision(2) << std::setw(12) << ratio << "\n";
}

}  // namespace

int main() {
    using namespace derivkit;

    const VanillaSpec spec{
        .spot = 100.0,
        .strike = 100.0,
        .rate = 0.05,
        .dividend = 0.0,
        .vol = 0.20,
        .time = 1.0,
        .type = OptionType::Call,
    };

    const double bs = bs::price(spec);
    mc::McConfig cfg{.paths = 50000, .seed = 42};

    std::cout << "European call, 50,000 paths, seed=42.  Black-Scholes = " << std::fixed
              << std::setprecision(6) << bs << "\n";
    std::cout << "Variance ratio = Var(crude) / Var(method)  (higher is better).\n\n";
    std::cout << std::left << std::setw(28) << "method" << std::right << std::setw(12) << "price"
              << std::setw(12) << "stderr" << std::setw(12) << "var ratio"
              << "\n";

    cfg.vr = VarianceReduction::None;
    const auto crude = mc::european(spec, cfg);
    row("crude", crude, crude.error_estimate);

    cfg.vr = VarianceReduction::Antithetic;
    row("antithetic", mc::european(spec, cfg), crude.error_estimate);

    cfg.vr = VarianceReduction::ControlVariate;
    row("control (S_T)", mc::european(spec, cfg), crude.error_estimate);

    cfg.vr = VarianceReduction::Antithetic | VarianceReduction::ControlVariate;
    row("antithetic + control", mc::european(spec, cfg), crude.error_estimate);

    std::cout << "\nArithmetic Asian call, 50 fixings, 20,000 paths.\n";
    std::cout << "Geometric Asian (Kemna-Vorst) = " << std::fixed << std::setprecision(6)
              << bs::geometric_asian(spec, 50) << "\n\n";

    mc::AsianConfig asian;
    asian.mc.paths = 20000;
    asian.mc.seed = 42;
    asian.steps = 50;

    asian.mc.vr = VarianceReduction::None;
    const auto a_crude = mc::arithmetic_asian(spec, asian);
    row("asian crude", a_crude, a_crude.error_estimate);

    asian.mc.vr = VarianceReduction::Antithetic;
    row("asian antithetic", mc::arithmetic_asian(spec, asian), a_crude.error_estimate);

    asian.mc.vr = VarianceReduction::ControlVariate;
    row("asian geo-control", mc::arithmetic_asian(spec, asian), a_crude.error_estimate);

    asian.mc.vr = VarianceReduction::Antithetic | VarianceReduction::ControlVariate;
    row("asian anti+geo-cv", mc::arithmetic_asian(spec, asian), a_crude.error_estimate);

    return 0;
}

File: src/derivkit/cli/variance_reduction.py
from __future__ import annotations

from derivkit.black_scholes import geometric_asian, price
from derivkit.monte_carlo import AsianConfig, McConfig, arithmetic_asian, european
from derivkit.result import PricingResult
from derivkit.types import OptionType, VanillaSpec, VarianceReduction


def _row(name: str, r: PricingResult, ref_err: float) -> None:
    ratio = (ref_err * ref_err) / (r.error_estimate * r.error_estimate) if ref_err > 0.0 else 0.0
    print(f"{name:<28}{r.value:12.6f}{r.error_estimate:12.3e}{ratio:12.2f}")


def main() -> int:
    spec = VanillaSpec(
        spot=100.0,
        strike=100.0,
        rate=0.05,
        dividend=0.0,
        vol=0.20,
        time=1.0,
        type=OptionType.CALL,
    )
    bs = price(spec)
    cfg = McConfig(paths=50000, seed=42)
    print(f"European call, 50,000 paths, seed=42.  Black-Scholes = {bs:.6f}")
    print("Variance ratio = Var(crude) / Var(method)  (higher is better).\n")
    print(f"{'method':<28}{'price':>12}{'stderr':>12}{'var ratio':>12}")

    cfg.vr = VarianceReduction.NONE
    crude = european(spec, cfg)
    _row("crude", crude, crude.error_estimate)
    cfg.vr = VarianceReduction.ANTITHETIC
    _row("antithetic", european(spec, cfg), crude.error_estimate)
    cfg.vr = VarianceReduction.CONTROL_VARIATE
    _row("control (S_T)", european(spec, cfg), crude.error_estimate)
    cfg.vr = VarianceReduction.ANTITHETIC | VarianceReduction.CONTROL_VARIATE
    _row("antithetic + control", european(spec, cfg), crude.error_estimate)

    print("\nArithmetic Asian call, 50 fixings, 20,000 paths.")
    print(f"Geometric Asian (Kemna-Vorst) = {geometric_asian(spec, 50):.6f}\n")

    asian = AsianConfig(mc=McConfig(paths=20000, seed=42), steps=50)
    asian.mc.vr = VarianceReduction.NONE
    a_crude = arithmetic_asian(spec, asian)
    _row("asian crude", a_crude, a_crude.error_estimate)
    asian.mc.vr = VarianceReduction.ANTITHETIC
    _row("asian antithetic", arithmetic_asian(spec, asian), a_crude.error_estimate)
    asian.mc.vr = VarianceReduction.CONTROL_VARIATE
    _row("asian geo-control", arithmetic_asian(spec, asian), a_crude.error_estimate)
    asian.mc.vr = VarianceReduction.ANTITHETIC | VarianceReduction.CONTROL_VARIATE
    _row("asian anti+geo-cv", arithmetic_asian(spec, asian), a_crude.error_estimate)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
